### ISSUE TITLE
Sierra Compatibility Dev Discussion

### ISSUE THREAD

**Issue by joecridge** [tekezo/Karabiner#739]
*Disclaimer: I have contributed absolutely nothing to this project. I’m just an interloper.*

**What we know:**

- Karabiner doesn’t work on macOS 10.12 Sierra.

**What @takezo has said:**

- > macOS 10.12 changes of the generic keyboard and mouse drivers made a
  > great impact on Karabiner and Seil.
  > We should make a large changes in Karabiner and Seil architecture.
  > There is not a workaround for this issue.
  >
  > Please wait an update of Karabiner and Seil for macOS 10.12.
  > (It may take a long time.)

- > I'm working on prototying the next generation of Karabiner and Seil but I
  > have to solve many problems.
  > https://github.com/tekezo/Karabiner-Elements

**Relevant issues:**

- #660 *macOS (10.12) compatibility*

- #711 *Karabiner is not working on sierra!!! Changed delete key to forward delete and assigned '`' key for forward delete*

- #732 *It's not working with MacOS Sierra*

**What we can do:**

- [Donate](https://pqrs.org/osx/karabiner/donation.html.en).

- Use the reaction buttons below to show your interest.

- Subscribe to this issue in the hope that it will contain **developer updates only.**

- Try [Karabiner Elements](https://github.com/tekezo/Karabiner-Elements/).
  If you were only doing simple things with Karabiner, it might be able to tide you over.

**What NOT to do:**

- Make a comment saying that you’ve donated. @takezo will know anyway.

- Make a comment asking how to do something in Elements, or reporting that a feature is missing. Use the Elements [issue tracker](https://github.com/tekezo/Karabiner-Elements/issues) (for developer issues) and Karabiner [Google group](https://groups.google.com/forum/#!forum/osx-karabiner) (for help/discussion) instead.

- Reply with a comment that just says ”+1”.

- Reply with any comment that you know won’t interest the subscribers of this issue, or help with Karabiner development.

Lastly, if you want to get really meta and discuss what’s written here, reply to me on #660 (since it’s already spammy there…) so that we can keep this issue tidy.

Cheers!

**Comment by greenhobby** [comment]
Just a note here that I'm running Sierra and it's working great. I have successfully made it swap the fn and ctrl keys on my Magic keyboard. There may be issues, but this is huge for my needs (an emacs user).

thank you soooooo much!!!!

**Comment by pkamb** [comment]
I've opened rdar://29426145 to hopefully give Apple some insight into the breaking keyboard driver changes that were introduced in Sierra.

Please feel free to **duplicate this radar** by filing a Bug Report.

https://bugreport.apple.com/

It's possible that Apple could introduce a compatibility layer in a future OS X update that would allow Karabiner (and other legacy keyboard apps) to function in macOS 10.12+.

["Filing duplicates is critical to getting things fixed."](https://blackpixel.com/writing/2012/02/radar-or-gtfo.html)

---

### Title:
macOS 10.12 Sierra keyboard driver changes have broken widely used Keyboard customization app

### Summary:
Opening this radar for Apple visibility.
Duplicate of rdar://29426145

Many Mac OS X users have installed the open-source keyboard customization utility "Karabiner" to remap their keyboards, provide internationalization support, improve keyboard accessibility, and otherwise improve their Mac keyboard experience.

The macOS 10.12 Sierra update has caused Karabiner to stop working. 

[https://github.com/tekezo/Karabiner/issues/739
](https://github.com/tekezo/Karabiner/issues/739)

The developer has said:

 > macOS 10.12 changes to the generic keyboard and mouse drivers had a great impact on Karabiner
 > These will require large changes to the Karabiner architecture.
 > There is not a workaround for this issue.

 > Please wait an update to Karabiner for macOS 10.12.
 > (It may take a long time.)

It's unclear if this app will ever be updated, or if the app's behavior will ever be possible under the new keyboard drivers.

If possible, a partial rollback or compatibility layer to support the legacy keyboard drivers would be greatly appreciated!

### Steps to reproduce:
Use open-source Keyboard customization app Karabiner in macOS Sierra. Karabiner uses pre-10.12 keyboard drivers.

### Expected Results:
 - Can use app Karabiner (which uses pre-10.12 keyboard drivers) to remap keyboard in Sierra.
 - Users can stay productive when using the latest release of macOS.

### Actual Results:
 - App Karabiner (which uses pre-10.12 keyboard drivers) does not function under macOS 10.12 due to system updates to keyboard drivers.
 - User cannot customize keyboard in Sierra.
 - User will not update to Sierra, and stay on macOS 10.11 instead, in order to keep their productive keyboard customizations.

### Additional Notes:
Many users have come to rely on this app to customize their keyboard. The lack of keyboard customization in Sierra is causing people not to update to the latest version of OS X.

From Apple, a legacy keyboard driver compatibility layer, partial rollback, migration document, or any other action with regards to these drivers would be greatly appreciated!

**Comment by idvorkin** [comment]
is there an ETA for when Karabiner will be supported on Sierra? Karabiner is so important for me that it's blocking my ability to upgrade to Sierra.

**Comment by makerofthings7** [comment]
Posted issue rdar://30413783

**Comment by jcrben** [comment]
How exactly does one look up these radar posts? Most openradar bugs have links like this: https://openradar.appspot.com/324324325

If I type in https://openradar.appspot.com/30413783 (presumably the one posted by @makerofthings7) I get nothing. Same with https://openradar.appspot.com/29426145

Just post HTTP links please...

**Comment by joecridge** [comment]
@jcrben I don’t believe that they are publicly accessible. You can see bugs that you’ve reported, and the *status* of any linked bugs, but that seems to be it.

**Comment by madmod** [comment]
Mine: [rdar://31042108](https://openradar.appspot.com/radar/31042108)

I also added that newer Apple hardware has no clear path to resolution as it is very difficult to install an older macOS on them and hardware may not be supported.

This is amazingly inconvenient for me and I hope it is fixed soon.

**Comment by pkamb** [comment]
> From Apple, a legacy keyboard driver compatibility layer, partial rollback, **migration document**, or any other action with regards to these drivers would be greatly appreciated!

I just saw a link to Apple's TN2450, which I had not ever seen mentioned in this Sierra upgrade discussion. Linking for visibility.

 >**Technical Note TN2450** - Remapping Keys in macOS 10.12 Sierra
 >
 > Under macOS Sierra 10.12, the mechanism for key remapping was changed. This Technical Note is for developers of key remapping software so that they can update their software to support macOS Sierra 10.12. We present 2 solutions for implementing key remapping functionality for macOS 10.12 in this Technical Note.
 > 
 > https://developer.apple.com/library/archive/technotes/tn2450/_index.html
